### DISCUSSION TITLE
Jupyter notebook interactive startup file

### DISCUSSION THREAD

**Discussion by tfabbri** [microsoft/pylance-release#3988] [Enhancement]
In my daily notebook development, I exploit the ipython startup file to automatically import packages/modules/functions to avoid annoying import lines every time I create a new notebook (or I run an ipython console). This feature works perfectly into vscode. 

The startup file that I am referring is located into `~/.ipython/profile_default/startup/start.py`. In my case, it contains the the definition of a routine like `addModulePath(...)` which modifies  he ` system.path` to allow the import of new modules easily. 

Unfortunately, pylance intellinsense, during notebook development, it is not able to recognize already loaded modules by `start.py` and modules added dynamically to the `system.path` - prompting to the user  messages like `addModulePath is not defined (PyLance) reportUndefinedVariable`  or for modules added to the system.path `Import package could not be resolved Pylance (reportMissingImport)`.  Example given in the image below:

![Screenshot from 2022-09-17 12-01-41](https://user-images.githubusercontent.com/3789256/190855610-29084c97-5c03-4506-8252-fd7b1a8c314a.png)

I have tried to modify the preferences by adding the `~/.ipython/profile_default/startup` to the Python Analysis: Extra Paths or Pythjon Analysis Include without any results. 

How to solve this? Is there a particular setup to solve this?

**Comment by jaymegordo** [comment]
Same request here, I've mentioned it before
https://github.com/microsoft/vscode-jupyter/issues/7098

**Comment by debonte** [comment]
> I have tried to modify the preferences by adding the ~/.ipython/profile_default/startup to the Python Analysis: Extra Paths or Pythjon Analysis Include without any results.
>
> How to solve this? Is there a particular setup to solve this?

We don't support `~/...` paths properly. I created a separate issue for this: https://github.com/microsoft/pylance-release/issues/3730

`extraPaths` will work if you use a path starting with `/home/myusername/` like so:
```json
"python.analysis.extraPaths": [
    "/home/myusername/.ipython/profile_default/startup"
]
```

With this configuration you'll be able to import symbols from your startup files. But I believe your goal is to have these symbols available without an explicit import. That is not possible today because Pylance has no knowledge of the `~/.ipython/profile_default/startup` directory. We could certainly add this functionality as an enhancement though.

In the meantime, for imports, variable declarations, and the like, would `jupyter.runStartupCommands` work for you? It's a VS Code setting that provides code (as a string or array) to be run before the first cell of all Jupyter notebooks. Pylance's support for `runStartupCommands` was broken when we switched to LSP-based notebooks, but it was recently [fixed](https://github.com/microsoft/pylance-release/issues/3219) in version [2022.11.41](https://github.com/microsoft/pylance-release/releases/tag/2022.11.41).

![image](https://user-images.githubusercontent.com/12820898/206069404-e90531b7-1b1b-4c33-80a5-f322827dd91f.png)

I realize that this approach has some limitations, but perhaps it's sufficient as a workaround?

> modules added dynamically to the `system.path`

@tfabbri, Pylance is a static analysis tool and therefore doesn't handle dynamic changes to system paths. You'll need to avoid this approach if you want to use a static type checker.

**Comment by tfabbri** [comment]
Sorry for getting back only now. 

@debonte 

> With this configuration you'll be able to import symbols from your startup files. But I believe your goal is to have these symbols available without an explicit import. That is not possible today because Pylance has no knowledge of the ~/.ipython/profile_default/startup directory. We could certainly add this functionality as an enhancement though.

Right, I think adding this functionality would be helpful. 

> In the meantime, for imports, variable declarations, and the like, would `jupyter.runStartupCommands` work for you?  ...
> I realize that this approach has some limitations, but perhaps it's sufficient as a workaround?

Thank you for the hint! I will consider this workaround, but I would prefer a leaner solution to be able to change configuration from one notebook from the other. If I will found an alternative workaround I will post it here. 
Thank you again for the precious support!

**Comment by debonte** [comment]
@tfabbri, thanks for your response. I've converted the issue to an enhancement discussion. If this gets sufficient up votes we'll consider implementing support for IPython startup files.
